src/Services/OdkLinkService.php

```diff
@@ -726,35 +724,132 @@ private function storeMainSurveyToCustomTable(Xlsform $xlsform, $entry, XlsformT
+
+            // create a new database record
+            $class::create($dataArray);
+        }
+    }
+
 
-            // Link the new data model to the current submission
-            $dataArray['submission_id'] = $submissionId;
+    // a generic function to extract values from main survey and repeat group entry, returns an array for further processing
+    private function prepareDataArray($xlsform, $entry, $section, $schema, $model, $submissionId): array
+    {
+        // initialise array
+        $result = [];
+
+        // add submission Id to array
+        $result['submission_id'] = $submissionId;
+
+        // get all column names of a table
+        $columnNames = Schema::getColumnListing($model->getTable());
+
+        // get all foreign key details of a table
+        $foreignKeyDetails = Schema::getForeignKeys($model->getTable());
+
+        // store foreign key column name and foreign key table name in associative array
+        // TODO: find Laravel array helper function to do the same in a simpler way
+        $foreignKeyColumnNames = [];
+        foreach ($foreignKeyDetails as $foreignKey) {
+            foreach ($foreignKey['columns'] as $foreignKeyColumn) {
+                $foreignKeyColumnNames[$foreignKeyColumn] = $foreignKey['foreign_table'];
+            }
+        }
 
-            // access the value of each ODK variable from a deeply nested array using "dot" notation
-            foreach ($schema as $schemaItem) {
+
+        // access the value of each ODK variable from a deeply nested array using "dot" notation
+        foreach ($schema as $schemaItem) {
+
+            // extract value from main survey
+            if ($section->is_repeat == 0) {
                 $itemPath = 'root' . Str::replace('/', '.', $schemaItem['path']);
                 $value = Arr::get($entry, $itemPath);
 
+                // extract value from repeat group
+            } else {
+                $pathLength = Str::length($schemaItem['path']);
+                $position = Str::position($schemaItem['path'], $section->structure_item);
+                $lengthToCut = $pathLength - $position;
+
+                $itemPath = Str::substr($schemaItem['path'], $position + Str::length($section->structure_item), $lengthToCut);
+                // dump('$itemPath : ' . $itemPath);
+
+                $fullItemPath = 'rg' . Str::replace('/', '.', $itemPath);
+                // dump('$fullItemPath : ' . $fullItemPath);
+
+                $value = Arr::get($entry, $fullItemPath);
                 // dump($schemaItem['name'] . ' : ' . $value);
+            }
+
+
+            // if app developer has defined a method of creating foreign key record in submission content, call that method:
+            $class = config('filament-odk-link.submission.foreign_key_process_method.class');
+            $method = config('filament-odk-link.submission.foreign_key_process_method.method');
```

Comment:
I think this works for now. Probably we should eventually move towards a set of interfaces / traits that give a clearer set of requirements to developers (and with the benefit that IDEs can read interfaces and tell you when you have missed something important.

Reply:
Yes, this would be cleaner and easier for developer to follow.